src/tactic/positivity.lean

```diff
@@ -297,18 +297,42 @@ private lemma div_nonneg_of_nonneg_of_pos [linear_ordered_field R] {a b : R} (ha
   0 ≤ a / b :=
 div_nonneg ha hb.le
 
+private lemma int_div_self_pos {a : ℤ} (ha : 0 < a) : 0 < a / a :=
+by { rw int.div_self ha.ne', exact zero_lt_one }
+
+private lemma int_div_nonneg {a b : ℤ} (ha : 0 ≤ a) (hb : 0 ≤ b) : 0 ≤ a / b := int.div_nonneg ha hb
+
+private lemma int_div_nonneg_of_pos_of_nonneg {a b : ℤ} (ha : 0 < a) (hb : 0 ≤ b) : 0 ≤ a / b :=
+int.div_nonneg ha.le hb
+
+private lemma int_div_nonneg_of_nonneg_of_pos {a b : ℤ} (ha : 0 ≤ a) (hb : 0 < b) : 0 ≤ a / b :=
+int.div_nonneg ha hb.le
+
+private lemma int_div_nonneg_of_pos_of_pos {a b : ℤ} (ha : 0 < a) (hb : 0 < b) : 0 ≤ a / b :=
+int.div_nonneg ha.le hb.le
+
 /-- Extension for the `positivity` tactic: division is nonnegative if both numerator and denominator
 are nonnegative, and strictly positive if both numerator and denominator are. -/
 @[positivity]
 meta def positivity_div : expr → tactic strictness
-| `(%%a / %%b) := do -- TODO handle eg `int.div_nonneg`
+| `(%%a / %%b) := do
   strictness_a ← core a,
   strictness_b ← core b,
-  match strictness_a, strictness_b with
-  | (positive pa), (positive pb) := positive <$> mk_app ``div_pos [pa, pb]
-  | (positive pa), (nonnegative pb) := nonnegative <$> mk_app ``div_nonneg_of_pos_of_nonneg [pa, pb]
-  | (nonnegative pa), (positive pb) := nonnegative <$> mk_app ``div_nonneg_of_nonneg_of_pos [pa, pb]
-  | (nonnegative pa), (nonnegative pb) := nonnegative <$> mk_app ``div_nonneg [pa, pb]
+  typ ← infer_type a,
+  match typ, strictness_a, strictness_b with
+  | `(ℤ), positive pa, positive pb :=
+      if a = b -- Only attempts to prove `0 < a / a`, otherwise falls back to `0 ≤ a / b`
+      then positive <$> mk_app ``int_div_self_pos [pa]
+      else nonnegative <$> mk_app ``int_div_nonneg_of_pos_of_pos [pa, pb]
```

Comment:
I believe you can get the same effect from
```suggestion
      positive <$> mk_app ``int_div_self_pos [pa]
      <|> nonnegative <$> mk_app ``int_div_nonneg_of_pos_of_pos [pa, pb]
```

Reply:
I doubt so because `positive <$> mk_app ``int_div_self_pos [pa]` will never fail. It's just constructing the proof term. Instead, the error will happen higher up in one of the core `positivity` tactics.

Am I understanding correctly, @digama0?

Reply:
Ah, I see.  One other comment here: elsewhere in this file I've been using the indentation convention
```suggestion
      if a = b then -- Only attempts to prove `0 < a / a`, otherwise falls back to `0 ≤ a / b`
        positive <$> mk_app ``int_div_self_pos [pa]
      else
       nonnegative <$> mk_app ``int_div_nonneg_of_pos_of_pos [pa, pb]
```

Reply:
@YaelDillies No, `mk_app` can fail if the required typeclasses are not found. Indeed I think that is the right solution in this case, assuming `int_div_self_pos` and `int_div_nonneg_of_pos_of_pos` have different typeclass assumptions such that the second can succeed while the first fails.

Reply:
`int_div_self_pos` and `int_div_nonneg_of_pos_of_pos` have no typeclass assumptions, though.